### PR TITLE
Work around memory issue with jruby/jruby#5148

### PR DESCRIPTION
**Pull request by jsvd**
on JRuby 9k, a read from the UDP socket will not truncate unused buffer space so we have to force it using String#s

Memory profile without this patch:

screen shot 2018-05-02 at 15 46 18

With the patch:

screen shot 2018-05-02 at 15 46 06

fixes https://github.com/elastic/logstash/issues/9346

### CODE REVIEW

**Comment by marioluan** [comment]
@jsvd Out of curiosity, how are you running the memory profile?

**Comment by jsvd** [comment]
@marioluan this is just yourkit.
My test is simply running:

```
 bin/logstash -e "input { udp { port => 4444 } } output { stdout { codec => dots } }"| pv -War > /dev/null
```

and a ruby script producing UDP data:

```ruby
require 'socket'
socket = UDPSocket.new
socket.connect("localhost", 4444)
loop { socket.send("test"*1000, 0) }
```

**Comment by jsvd** [comment]
one thing we could do is use the charset that is defined in the codec for the plugin

**Comment by jsvd** [comment]
interestingly, it seems to work with utf8 messages.

I tried `echo -n "にほんご" | nc -u 127.0.0.1 4444`, and got on logstash:

```
% bin/logstash -e "input { udp { port => 4444 } } output { stdout { codec => rubydebug } }"
{
      "@version" => "1",
       "message" => "にほんご",
    "@timestamp" => 2018-05-02T17:03:09.640Z,
          "host" => "127.0.0.1"
}
```

**Comment by original-brownbear** [comment]
@andrewvc taking a look now, will answer shortly. Sorry for the delay here!

**Comment by original-brownbear** [comment]
@andrewvc I'm not so sure this approach will work with the PQ (or in general). We're def. loosing information about the initial encoding here. 

See:

```sh
jruby-9.1.13.0 :005 > "ÄÖÜP".b.to_java.to_string
 => "Ã\u0084Ã\u0096Ã\u009CP" 
jruby-9.1.13.0 :006 > "ÄÖÜP".to_java.to_string
 => "ÄÖÜP" 
```

=> we def. break PQ serialisation of these strings :(

**Comment by original-brownbear** [comment]
@andrewvc I think the trick here is to simply use `.b` to truncate the internal buffers and then re-encode to the original encoding maybe?

**Comment by webmat** [comment]
@jsvd As I already mentioned, I'm not familiar enough with all of the tooling yet to confidently test the memory usage of the plugin itself.

**Memory**

In my own contrived test (plain UDP socket in pry), I clearly observed the memory blowing up with YourKit, using your code example in the JRuby issue you opened. Modifying said code (see below) with a fix similar to your PR, I could no longer observe memory blowing up with YourKit.

```ruby
data = []
puts "Listening for UDP data"
consumer = Thread.new do
  while true do
    begin
      payload, client = udp.recvfrom_nonblock(65536)
      next if payload.empty?
      data << payload.b.force_encoding(Encoding::UTF_8)
    rescue IO::EAGAINWaitReadable
      sleep 1
    end
  end
end
```

So this seems to fix the memory issue itself.

**Encoding**

When looking at the payload encoding right after receiving it, its encoding is actually ASCII-8BIT, not UTF-8. On the other hand UTF-8 is a superset of ASCII, and the message ends up being converted to UTF-8 anyway with 3.3.2, prior to your PR, so this doesn't seem like a big deal.

The following hack is *based on v3.3.2, not your PR*:

```ruby
          payload, client = @udp.recvfrom_nonblock(@buffer_size)
          puts payload.encoding.inspect
          break if payload.empty?
```

Then from inside Logstash, I look at the encoding of the message with a ruby filter.

```
bin/logstash -e 'input { udp { port => 4444 } } 
                filter { ruby { code => "event.set( %[encoding], event.get(%[message]).encoding.inspect )" } }
                output { stdout { codec => rubydebug } }'
```

When running this and sending a message with netcat, here's the output I get:

```
[2018-05-08T15:33:10,766][INFO ][logstash.inputs.udp      ] UDP listener started {:address=>"0.0.0.0:4444", :receive_buffer_bytes=>"196724", :queue_size=>"2000"}
#<Encoding:ASCII-8BIT>
{
          "host" => "127.0.0.1",
      "encoding" => "#<Encoding:UTF-8>",
       "message" => "ohai\n",
      "@version" => "1",
    "@timestamp" => 2018-05-08T19:33:17.277Z
}
```

So since the message already ends up as UTF-8 anyway, I guess the enforce_encoding is fine.

TL;DR: LGTM ;-)

But if you want validation on the memory part, make sure to loop in someone else better versed with YourKit. This was my first time using it, and it always crashed my Logstash.

**Comment by jsvd** [comment]
Thanks for the thorough explanation and review @webmat , I think we're good here.

WRT the encoding, I guess we could remove the force to utf8, but let's be extra safe.
